roachtest: add groups to task manager

herkolategan opened the pull request:
Previously, a new API for managing tasks were introduced (see #133263). However, this did not address roachtests that want to manage groups. In an effort to replace `errgroup`, and `monitor.Go` for task management, this change introduces a group provider in the task manager.

A group adds the ability to wait on a subset of tasks to finish before proceeding. The task handler will still report returned errors or panics to the test framework.

Informs: #118214

Epic: None
Release note: None

herkolategan committed:
roachtest: add groups to task manager

Previously, a new API for managing tasks were introduced (see #133263). However,
this did not address roachtests that want to manage groups. In an effort to replace
`errgroup`, and possibly using `monitor.Go` for task management this change
introduces a group provider in the task manager.

A group adds the ability to wait on a subset of tasks to finish before
proceeding. The task handler will still report returned errors or panics to the
test framework.

Informs: #118214

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/135831)
<!-- Reviewable:end -->

herkolategan commented on pkg/cmd/roachtest/roachtestutil/task/manager_test.go:
```diff
@@ -135,6 +135,53 @@ func TestTerminate(t *testing.T) {
+
+	select {
+	case <-done:
+	case <-time.After(30 * time.Second):
```
I'm curious if this is good practice to have a timeout in a test, or rather wait indefinitely and let a "global test timeout" happen?

DarrylWong replied:
I'm not sure how stress/race interacts with time, but it does feel like this would break under that.

herkolategan committed:
roachtest: add groups to mvt helper

This change groups the default options used by tasks and groups into a function
that can be shared when creating new groups or tasks via the mixedversion test
helper. The mixedversion helper now also implements `GroupProvider` to enable
grouping tasks.

Informs: #118214

Epic: None
Release note: None

herkolategan committed:
roachtest: add groups to test framework

This change groups the default options used by tasks and groups into a function
that can be shared when creating new groups or tasks via the test framework. The
`Test` interface now implements `GroupProvider` as well to enable grouping
tasks.

Informs: #118214

Epic: None
Release note: None

herkolategan committed:
roachtest: use a task group in multitenant upgrade test

Previously, a roachtest would need to use wait groups in order to wait on tasks.
The new `GroupProvider` interface now allows tests to manage groups of tasks via
the framework instead of manually.

Informs: #118214

Epic: None
Release note: None

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/task/manager.go:
```diff
@@ -73,9 +87,62 @@ func (m *manager) defaultOptions() []Option {
+	_ = m.group.GoWithCancel(fn, opts...)
+}
+
+func (t *group) NewGroup(opts ...Option) Group {
```
For example, in the new unit test you added, if instead of closing the channels for each task you called `manger.Terminate()`, I would expect all the group tasks to get cancelled and unblock.

```
diff --git a/pkg/cmd/roachtest/roachtestutil/task/manager_test.go b/pkg/cmd/roachtest/roachtestutil/task/manager_test.go
index 54fe3838313..c7baf8a71de 100644
--- a/pkg/cmd/roachtest/roachtestutil/task/manager_test.go
+++ b/pkg/cmd/roachtest/roachtestutil/task/manager_test.go
@@ -148,7 +148,10 @@ func TestGroups(t *testing.T) {
        for i := 0; i < numTasks; i++ {
                channels[i] = make(chan struct{})
                g.Go(func(ctx context.Context, l *logger.Logger) error {
-                       <-channels[i]
+                       select {
+                       case <-ctx.Done():
+                       case <-channels[i]:
+                       }
                        return nil
                })
        }
@@ -162,16 +165,19 @@ func TestGroups(t *testing.T) {
 
        // Close channels one by one to complete all tasks, and ensure the group is
        // not done yet.
-       for i := 0; i < numTasks; i++ {
-               select {
-               case <-done:
-                       t.Fatal("group should not be done yet")
-               default:
-               }
-               // Close the channel and wait for the completed event.
-               close(channels[i])
-               <-m.CompletedEvents()
-       }
+       m.Terminate(nilLogger())

```

herkolategan replied:
Nice catch!
😱 Massive oversight while refactoring and no unit test to catch it!
I added a unit test now, that will test exactly this case.

DarrylWong reviewed (commented):
Sorry for the slow review!

herkolategan commented on pkg/cmd/roachtest/roachtestutil/task/manager_test.go:
```diff
@@ -135,6 +135,53 @@ func TestTerminate(t *testing.T) {
+		<-m.CompletedEvents()
+	}
+
+	select {
```
I opted to remove the timeouts. A part of me does not like that a failure will then present as a package timeout, but I think as you mentioned adding a timeout here will probably mess with stress / race. In general there is no issue here as long as the tests are passing, but a failure won't be as obvious.

herkolategan commented:
Running a roachtest TC job with tests that has some `t.Go` calls in it for confirmation.

DarrylWong reviewed (approved):
LGTM assuming the test run looks okay.

herkolategan commented:
TFTR!

bors r=DarrylWong
